Merge ipv6 support

benjamreis opened the pull request.

benjamreis commented:
https://github.com/xenserver/host-installer/pull/2

stormi commented:
~~Looks good. As discussed, let's not create a merge commit when merging this PR. Fast-forward with git directly instead.~~ Not true anymore

ydirson commented on backend.py:
```diff
@@ -1535,15 +1535,15 @@ def configureNetworking(mounts, admin_iface, admin_bridge, admin_config, hn_conf
-                print >>mc, "DNS='%s'" % (','.join(nameservers),)
-            if domain:
-                print >>mc, "DOMAIN='%s'" % domain
         print >>mc, "MODEV6='%s'" % netinterface.NetInterface.getModeStr(admin_config.modev6)
         if admin_config.modev6 == netinterface.NetInterface.Static:
             print >>mc, "IPv6='%s'" % admin_config.ipv6addr
             if admin_config.ipv6_gateway:
                 print >>mc, "IPv6_GATEWAY='%s'" % admin_config.ipv6_gateway
+        if manual_nameservers:
+            print >>mc, "DNS='%s'" % (','.join(nameservers),)
+        if domain:
+            print >>mc, "DOMAIN='%s'" % domain
```
Would warrant to be a preparatory commit of its own, explaining that it changes nothing on IPv4 (and is it really the case, both in TUI and answerfile modes?), but will be needed for IPv6 (and why that structuring used for v4 does not fit v6)

ydirson commented on backend.py:
```diff
@@ -1585,12 +1585,17 @@ def configureNetworking(mounts, admin_iface, admin_bridge, admin_config, hn_conf
-    else:
-        nfd.write("NETWORKING_IPV6=no\n")
-        netutil.disable_ipv6_module(mounts["root"])
+    with open("%s/etc/sysctl.d/91-net-ipv6.conf" % mounts["root"], "w") as ipv6_conf:
```
This makes a big patch hunk, just to add 2 for/write loops that are already very similar.
Would be better to add a single helper function doing the with/for/write.

ydirson commented on netinterface.py:
```diff
@@ -16,7 +16,7 @@ def getTextOrNone(nodelist):
             rc = rc + node.data
     return rc == "" and None or rc.strip().encode()
 
-class NetInterface:
```
This is a python2 cleanup, should be in a separate preparatory commit (especially as the IPv6 support is unlikely to reach upstream before python3 support)

benjamreis replied:
It's to allow inheritance in the `Netinterfacev6` Object

ydirson commented on netinterface.py:
```diff
@@ -36,31 +36,25 @@ def __init__(self, mode, hwaddr, ipaddr=None, netmask=None, gateway=None,
-            assert netmask
+        is_static = mode == self.Static
+        if is_static:
+            assert ipaddr and netmask
```
keeping 2 separate asserts helps to know which condition was triggered

ydirson commented on netinterface.py:
```diff
@@ -36,31 +36,25 @@ def __init__(self, mode, hwaddr, ipaddr=None, netmask=None, gateway=None,
+        self.dns = dns if is_static else None
+        self.domain = domain if is_static else None
+        self.vlan = vlan
+
```
As a readability refactoring, this should live in its own preparatory commit

ydirson commented on netinterface.py:
```diff
@@ -124,7 +118,10 @@ def valid(self):
     def isStatic(self):
         """ Returns true if a static interface configuration is represented. """
-        return self.mode == self.Static
+        return self.mode == self.Static or (self.mode == None and self.modev6 == self.Static)
```
The sole use for `isStatic` is related to skipping an `arping` test, and `arping` only supports IPv4. Isn't it necessary to rework `waitUntilUp` instead?

benjamreis replied:
It is used in the TUI to choose to show or not buttons etc and as below, the IPv4 settings has priority.

ydirson commented on netinterface.py:
```diff
@@ -124,7 +118,10 @@ def valid(self):
+        return self.mode == self.Static or (self.mode == None and self.modev6 == self.Static)
+
+    def isDHCP(self):
+        return self.mode == self.DHCP or (self.mode == None and self.modev6 == self.DHCP)
```
So if we configure static IPv4 and DHCP for v6, `isDHCP()` is False, is it what we want?

benjamreis replied:
Yes IPv4 have priority by design to keep previous behaviour and as it will be the primary adress type of the host.

ydirson commented on netinterface.py:
```diff
@@ -143,13 +140,12 @@ def writeDebStyleInterface(self, iface, f):
         # IPv6 is only enabled through answerfiles and so is not supported here.
-        assert self.modev6 is None
-        assert self.mode
+        assert self.modev6 or self.mode
```
keeping 2 separate asserts helps to know which condition was triggered

benjamreis replied:
It's an or, we need to have one of the 2 configured, which is why None is now a valid value.

ydirson commented on netinterface.py:
```diff
@@ -143,13 +140,12 @@ def writeDebStyleInterface(self, iface, f):
         if self.mode == self.DHCP:
             f.write("iface %s inet dhcp\n" % iface_vlan)
-        else:
+        elif self.mode == self.Static:
```
`mode==None` was a already valid value, so this looks like a fix that can be applied before IPv6, seems to warrant its own commit too

benjamreis replied:
Since there's an `assert mode` mode can't be None here so i should let the simple else.

ydirson commented on netinterface.py:
```diff
@@ -160,12 +156,21 @@ def writeDebStyleInterface(self, iface, f):
+            f.write("iface %s inet6 dhcp\n" % iface_vlan)
+        if self.modev6 == self.Autoconf:
+            f.write("iface %s inet6 auto\n" % iface_vlan)
+        elif self.modev6 == self.Static:
```
consistency: either all `elif` or all `if`

ydirson commented on netinterface.py:
```diff
@@ -160,12 +156,21 @@ def writeDebStyleInterface(self, iface, f):
 
-        assert self.modev6 is None
-        assert self.mode
+        assert self.modev6 or self.mode
```
keeping 2 separate asserts helps to know which condition was triggered

benjamreis replied:
or here as well

ydirson commented on netinterface.py:
```diff
@@ -175,17 +180,38 @@ def writeRHStyleInterface(self, iface):
+        if self.mode == self.Static:
             # CA-11825: broadcast needs to be determined for non-standard networks
             bcast = self.getBroadcast()
-            f.write("BOOTPROTO=none\n")
```
part of the same `mode==None` fix suggested above?

benjamreis replied:
Same as above.

ydirson commented on netinterface.py:
```diff
@@ -25,7 +25,7 @@ class NetInterface:
     def __init__(self, mode, hwaddr, ipaddr=None, netmask=None, gateway=None,
                  dns=None, domain=None, vlan=None):
-        assert mode is None or mode == self.Static or mode == self.DHCP
+        assert mode in [None, self.Static, self.DHCP, self.Autoconf]
```
Distinct from the addition of a new value, there is a similar assert in `addIPv6` which would benefit from the same refactoring.  Would be worth having a separate preparatory commit to improve all of them.

ydirson commented on netinterface.py:
```diff
@@ -175,17 +180,38 @@ def writeRHStyleInterface(self, iface):
+        elif self.modev6 == self.Static:
+            f.write("IPV6ADDR=%s\n" % self.ipv6addr)
+            if self.ipv6_gateway:
+                f.write("IPV6_DEFAULTGW=%s\n" % (self.ipv6_gateway))
```
It seems awkward to have the `Autoconf` case handled that way.  It would be more visible that all cases are covered bo have it be a 3rd branch of that `if`, even if that means duplicating the `AUTOCONF=no` line for the other two cases. And we could then add `elif == None: pass else: assert()` to make the code future-robust.

ydirson commented on netinterface.py:
```diff
@@ -348,3 +374,21 @@ def loadFromNetDb(jdata, hwaddr):
+    def __init__(self, mode, hwaddr, ipaddr=None, netmask=None, gateway=None, dns=None, domain=None, vlan=None):
+        super(NetInterfaceV6, self).__init__(None, hwaddr, None, None, None, None, None, vlan)
+
+        is_static = mode == self.Static
```
`is_static` is not useful in this case

ydirson commented on netinterface.py:
```diff
@@ -348,3 +374,21 @@ def loadFromNetDb(jdata, hwaddr):
         nic.addIPv6(modev6, ipv6addr, gatewayv6)
         return nic
+
+class NetInterfaceV6(NetInterface):
```
The design underlying this class ought to be presented somewhere.  Since there is no design doc in this project, a good place could be the commit message.

ydirson reviewed (changes requested):
My first pass is limited to `backend` and `netinterface`.

As a more general note, I remember we previously talked about `writeDebStyleInterface()`, which looks like dead code. This PR makes it even more useless by leaving it only supporting IPv4, so I think it should first be removed in a preparatory commit.

benjamreis committed:
Write `manual_nameservers` and `domain` for IPv6 too

Both fields are used for the 2 IP families and were writen in
an IPv4 specific block.

Signed-off-by: BenjiReis <benjamin.reis@vates.fr>

benjamreis committed:
Add `NetInterfaceV6` to init an IPv6 interface

Inherits from `NetInterface` to mutualize the code

Signed-off-by: BenjiReis <benjamin.reis@vates.fr>

benjamreis committed:
isStatic helpers by IP family

Signed-off-by: BenjiReis <benjamin.reis@vates.fr>

benjamreis committed:
Use `socket.inet_pton` to validate IP

Add family specific validator as well

Signed-off-by: BenjiReis <benjamin.reis@vates.fr>

benjamreis committed:
Take IPv6 addresses into account to determine an interface is up

Signed-off-by: BenjiReis <benjamin.reis@vates.fr>

benjamreis committed:
Enable IPv6 in the host when it is configured

Signed-off-by: BenjiReis <benjamin.reis@vates.fr>

benjamreis committed:
Write IPv6 conf files for the installer

Signed-off-by: BenjiReis <benjamin.reis@vates.fr>

benjamreis committed:
Configure IPv6 in TUI

Add a screen to choose to configure IPv4, IPv6 or both
Display the interface screen for IPv4 and/or IPv6

Signed-off-by: BenjiReis <benjamin.reis@vates.fr>

benjamreis committed:
NetInterface inherits from Object

Signed-off-by: BenjiReis <benjamin.reis@vates.fr>

benjamreis committed:
`waitUntilUp` also wait for IPv6 if statically configured

Need to add `ndisc6` to the install.img

Signed-off-by: BenjiReis <benjamin.reis@vates.fr>

benjamreis committed:
Hide dynamic buttons in tui when in static mode

Add `isDynamic` helper to `NetInterface`

Signed-off-by: BenjiReis <benjamin.reis@vates.fr>

benjamreis committed:
Keep IPv6 enablement/disablement upon upgrades

Add `etc/sysctl.d/91-net-ipv6.conf` in the restore list

Signed-off-by: BenjiReis <benjamin.reis@vates.fr>

ydirson commented on backend.py:
```diff
@@ -1535,15 +1535,15 @@ def configureNetworking(mounts, admin_iface, admin_bridge, admin_config, hn_conf
-            if domain:
-                print >>mc, "DOMAIN='%s'" % domain
         print >>mc, "MODEV6='%s'" % netinterface.NetInterface.getModeStr(admin_config.modev6)
         if admin_config.modev6 == netinterface.NetInterface.Static:
```
(upstreaming)
commit message says "ipv6 too" but what I meant is that what it really does is writing them "also if not IPv4 static", so the needed explanation would rather be why it's OK for DHCPv4 (or for any non-static v6 mode)

ydirson commented on netutil.py:
```diff
@@ -137,7 +137,11 @@ def interfaceUp(interface):
+        return True
+
+    inet6s = filter(lambda x: x.startswith("    inet6 "), out.split("\n"))
+    return len(inet6s) > 1  # Not just the fe80:: address
```
(upstreaming)
the `> 1`  in v6 case make the original `== 1` for v4 look very suspicious, worth a preliminary commit to use `> 1` too?

benjamreis replied:
No, the `> 1` for IPv6 is to be sure that there's not only the link local ip `fe80:...` which is not relevant for IPv4.

ydirson commented on netinterface.py:
```diff
@@ -188,6 +187,27 @@ def writeRHStyleInterface(self, iface):
+            f.write("IPV6_DEFROUTE=yes\n")
+            f.write("IPV6_DEFAULTDEV=%s\n" % iface_vlan)
+
+        if self.modev6 == self.DHCP:
+            f.write("DHCPV6C=yes\n")
+            f.write("PERSISTENT_DHCLIENT_IPV6=yes\n")
+            f.write("IPV6_FORCE_ACCEPT_RA=yes\n")
+            f.write("IPV6_AUTOCONF=no\n")
+        elif self.modev6 == self.Static:
+            f.write("IPV6ADDR=%s\n" % self.ipv6addr)
+            if self.ipv6_gateway:
+                f.write("IPV6_DEFAULTGW=%s\n" % (self.ipv6_gateway))
+            f.write("IPV6_AUTOCONF=no\n")
+        elif self.modev6 == self.Autoconf:
+            f.write("IPV6_AUTOCONF=yes\n")
+
```
would be better inside `if self.modev6` for clarity?

ydirson commented on tui/network.py:
```diff
@@ -9,128 +9,215 @@
+
+        def dhcp_change():
+            for x in [ ip_field, gateway_field, subnet_field, dns_field ]:
+                x.setFlags(FLAG_DISABLED, static_rb.selected())
```
(upstreaming)
such changes are harder to spot with the reindent in the same commit, and `diff -w` is harder to read
would be more readable to split the introduction of the new function in a preliminary commit

ydirson commented on tui/network.py:
```diff
@@ -9,128 +9,215 @@
+        dns_field = Entry(16)
+        vlan_field = Entry(16)
+
+        static = bool(defaults and (defaults.modev6 if ipv6 else defaults.mode) == NetInterface.Static)
```
unnecessary cast to bool?

benjamreis replied:
if default is None then static is equal to None as the and is not evaluated.
And it can"t be passed as a bool in the following code.

ydirson commented on tui/network.py:
```diff
@@ -9,128 +9,215 @@
+        gf.add(entry_grid, 0, 4, padding=(0, 0, 0, 1))
+        if ipv6:
+            gf.add(autoconf_rb, 0, 5, anchorLeft=True)
+        # One more line for IPv6 autoconf
+        gf.add(vlan_cb, 0, 5 + ipv6, anchorLeft=True)
+        gf.add(vlan_grid, 0, 6 + ipv6, padding=(0, 0, 0, 1))
+        gf.add(buttons, 0, 7 + ipv6, growx=1)
```
the common idiom in the TUI is to leave a blank line rather than complicating layout computations

ydirson replied:
and `ipv6` is a `bool`, we should not rely on `5 + True` being `6`

ydirson commented on tui/network.py:
```diff
@@ -9,128 +9,215 @@
+        vlan_value = int(vlan_field.value()) if vlan_cb.selected() else None
+        if bool(dhcp_rb.selected()):
+            answers = iface_class(NetInterface.DHCP, nic.hwaddr, vlan=vlan_value)
+        elif ipv6 and bool(autoconf_rb.selected()):
```
bool unnecessary?

ydirson commented on tui/network.py:
```diff
@@ -286,23 +373,37 @@ def specify_configuration(answers, txt, defaults):
 
-        # check that we have *some* network:
-        if netutil.ifup(ifaceName) != 0 or not netutil.interfaceUp(ifaceName):
+        def display_error():
```
also reindenting in the same commit - all such new-func reindentations could live in a single separate commit

ydirson commented on netinterface.py:
```diff
@@ -214,13 +214,17 @@ def writeRHStyleInterface(self, iface):
+        if self.isStatic4() and self.gateway:
+            rc = util.runCmd2(['/usr/sbin/arping', '-f', '-w', '120', '-I',
+                           iface_name, self.gateway])
+        if rc != 0:
```
better inside previous `if`

ydirson commented on netinterface.py:
```diff
@@ -214,13 +214,17 @@ def writeRHStyleInterface(self, iface):
 
-        rc = util.runCmd2(['/usr/sbin/arping', '-f', '-w', '120', '-I',
-                           self.getInterfaceName(iface), self.gateway])
         return rc == 0
```
would likely be more clear/consistent to have another `if rc != 0:` right after `rc =` in the `if` block, and a plain `return True` to finish (even though one of the 2 if's have to be taken, it's not locally obvious that we ensure `rc` will be bound at this point)

ydirson commented on netutil.py:
```diff
@@ -92,7 +92,7 @@ def writeResolverFile(configuration, filename):
+        if (not settings.isDynamic()) and settings.dns:
             if settings.dns:
                 for server in settings.dns:
                     outfile.write("nameserver %s\n" % server)
```
The original code seems strange: why even check `isStatic` and not rely on `settings.dns` being previously set?  That seems to encode a single policy in 2 places.

It would be worth explaining in the commit message what the exact policy is: if one of v4|v6 is static but the other is dynamic, we don't get to select a DNS because we're necessarily happy with what the other dynamically-provided info is?

benjamreis replied:
The dynamic settings will override the statically configured.

stormi commented:
Looks like all the commits are on top of origin/10.10.5-8.3. Shouldn't there be a merge commit somewhere so that we can then fast-forward this branch to 10.10.5-8.3 as planned initially?
![image](https://github.com/xcp-ng/host-installer/assets/451043/c369fce7-6d21-4dab-8848-2a0af20142b6)

ydirson reviewed (commented):
Lots of little remarks (mostly) targeted to upstreaming branch

stormi commented:
We merge this current state for the next alpha/beta release of XCP-ng 8.3, and will continue improving the upstream PR to take the last non-blocking comments into account.
